java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -183,17 +201,50 @@ private void registerMinion(String minionId, boolean isSaltSSH, Optional<Long> s
         Opt.consume(reActivationKey,
             //Case-1 Registration
             () -> {
-                Optional<MinionServer> registeredMinionOpt = MinionServerFactory.findByMachineId(machineId);
-                Opt.consume(registeredMinionOpt,
-                    () -> {
-                        if (!duplicateMinionNamePresent(minionId)) {
-                            finalizeMinionRegistration(minionId, machineId, saltSSHProxyId, actKeyOverride, isSaltSSH);
+                Opt.consume(ServerFactory.findByMachineId(machineId), () -> {
+                    Opt.consume(MinionServerFactory.findByMinionId(minionId), () -> {
+                        // Case 1.1 - new registration
+                        finalizeMinionRegistration(minionId, machineId, saltSSHProxyId, actKeyOverride, isSaltSSH);
+                    }, minionServer -> {
+                        // Case 2.1 - update found system with new values
+                        LOG.warn(String.format(
+                                "A system with minion_id '%s' already exists, but with different " +
+                                        "machine-id ( %s vs. %s). Updating existing system with System ID: %s",
+                                minionId, machineId, minionServer.getMachineId(), minionServer.getId()));
+                        updateAlreadyRegisteredInfo(minionId, machineId, minionServer);
+                        applySaltBootStates(minionId, minionServer, saltbootInitrd);
+                    });
+                }, server -> {
+                    Opt.consume(MinionServerFactory.findByMinionId(minionId), () -> {
+                        // Case 1.2 - System got a new minion id
+                        Opt.consume(server.asMinionServer(),
+                                () -> {
+                                    // traditional client wants migration to salt
+                                    finalizeMinionRegistration(minionId, machineId, saltSSHProxyId, actKeyOverride,
+                                            isSaltSSH);
+                                },
+                                registeredMinion -> {
+                                    updateAlreadyRegisteredInfo(minionId, machineId, registeredMinion);
+                                    applySaltBootStates(minionId, registeredMinion, saltbootInitrd);
+                                });
+                    }, minionServer -> {
+                        // Case 2.2 - Cleanup missing - salt DB got out of sync with Uyuni DB
+                        // Can only happen when salt key was deleted and same minion id was accepted again
```

Comment:
Thinking about it, this comment is better to be below in `else` branch as the `if` branch is regular already registered minion startup.

Reply:
I will push another commit later which restructure it a bit. This will also change the place of this comment